src/main/java/org/elasticsearch/indices/IndicesService.java

```diff
@@ -419,4 +418,128 @@ public synchronized void beforeIndexShardClosed(ShardId shardId, @Nullable Index
             }
         }
     }
+
+    /**
+     * Deletes the given index. Persistent parts of the index
+     * like the shards files, state and transaction logs are removed once all resources are released.
+     *
+     * Equivalent to {@link #removeIndex(String, String)} but fires
+     * different lifecycle events to ensure pending resources of this index are immediately removed.
+     * @param index the index to delete
+     * @param reason the high level reason causing this delete
+     */
+    public void deleteIndex(String index, String reason) throws IOException {
+        removeIndex(index, reason, true);
+    }
+
+    public void deleteClosedIndex(String reason, ClusterState clusterState, IndexMetaData metaData) {
+        try {
+            deleteIndexStore(reason, clusterState, metaData);
+        } catch (IOException e) {
+            logger.warn("[{}] failed to delete closed index", e, metaData.index());
+        }
+    }
+
+    /**
+     * Deletes the index store trying to acquire all shards locks for this index.
+     * This method will delete the metadata for the index even if the actual shards can't be locked.
+     */
+    public void deleteIndexStore(String reason, ClusterState clusterState, IndexMetaData metaData) throws IOException {
+        if (nodeEnv.hasNodeFile()) {
+            String indexName = metaData.getIndex();
+            if (clusterState.metaData().hasIndex(indexName)) {
```

Comment:
I wonder if we should use the cluster state for this check. I'm worried about people passing in a dated cluster state here. Maybe a cleaner model is to make this method synchronised (to avoid async collision with the create code) and check the existence of the index instance in the #indices map member of this class.